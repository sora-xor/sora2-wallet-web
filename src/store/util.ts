import { mapGetters, mapState, mapActions, mapMutations } from 'vuex';
import { createDecorator, VueDecorator } from 'vue-class-component';

export enum VuexOperation {
  State = 'state',
  Getter = 'getter',
  Mutation = 'mutation',
  Action = 'action',
}

function getVuexMapFn(type: VuexOperation) {
  switch (type) {
    case VuexOperation.State:
      return mapState;
    case VuexOperation.Getter:
      return mapGetters;
    case VuexOperation.Mutation:
      return mapMutations;
    case VuexOperation.Action:
      return mapActions;
  }
}

function isComputedDecorator(type: VuexOperation): boolean {
  return [VuexOperation.State, VuexOperation.Getter].includes(type);
}

export function attachDecorator(type: VuexOperation, modulesChain: string, name: string): VueDecorator {
  const mapFn = getVuexMapFn(type);
  const featuresType = isComputedDecorator(type) ? 'computed' : 'methods';
  return createDecorator((options, key) => {
    let features = { ...options[featuresType] };
    if (!features) {
      features = {};
    }
    const mapObject = { [name]: name };
    features[key] = mapFn(modulesChain, mapObject)[name];
    options[featuresType] = features;
  });
}

export function createDecoratorsObject(
  obj: any,
  newObj: any,
  modules: Array<string>,
  type: VuexOperation,
  path?: string
): void {
  let keys = Object.keys(obj);
  if (!keys.length) {
    // cuz Object.keys refers only to enumerable props
    keys = Object.getOwnPropertyNames(obj);
  }
  for (const key of keys) {
    const value = obj[key];
    if (path && modules.includes(path)) {
      newObj[key] = attachDecorator(type, path, key);
    } else {
      newObj[key] = {};
      createDecoratorsObject(value, newObj[key], modules, type, path ? `${path}/${key}` : key);
    }
  }
}
