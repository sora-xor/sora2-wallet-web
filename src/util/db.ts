import { openDB, DBSchema } from 'idb';

interface MetadataDB extends DBSchema {
  metadata: {
    value: {
      genesisHash: string;
      specVersion: number;
      metadata: string;
    };
    key: string;
  };
}

const METADATA_DB_NAME = 'metadata';
const METADATA_DB_VERSION = 1;
const METADATA_DB_CONNECTION = openDB<MetadataDB>(METADATA_DB_NAME, METADATA_DB_VERSION, {
  upgrade(db) {
    db.createObjectStore('metadata', {
      keyPath: 'genesisHash',
    });
  },
});

export async function getMetadataStore() {
  if (!('indexedDB' in window)) {
    return {
      get: undefined,
      set: undefined,
    };
  }

  const db = await METADATA_DB_CONNECTION;

  const get = async (genesisHash: string, specVersion: number): Promise<string | null> => {
    const time1 = performance.now();
    const record = await db.get('metadata', genesisHash);

    if (record) {
      if (record.specVersion === specVersion) {
        console.log('get', performance.now() - time1);
        return record.metadata;
      } else {
        await db.delete('metadata', genesisHash);
      }
    }

    console.log('get', performance.now() - time1);

    return null;
  };

  const set = async (genesisHash: string, specVersion: number, metadata: string) => {
    const time1 = performance.now();
    await db.put('metadata', { genesisHash, specVersion, metadata });
    console.log('set', performance.now() - time1);
  };

  return { get, set };
}
