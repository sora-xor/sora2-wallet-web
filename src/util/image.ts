import { saveAs } from 'file-saver';
import base64 from 'base-64';

export enum IMAGE_EXTENSIONS {
  SVG = '.svg',
  PNG = '.png',
  JPEG = '.jpeg',
  WEBP = '.webp',
  GIF = '.gif',
}

export const IMAGE_MIME_TYPES = {
  [IMAGE_EXTENSIONS.SVG]: 'image/svg+xml',
  [IMAGE_EXTENSIONS.PNG]: 'image/png',
  [IMAGE_EXTENSIONS.JPEG]: 'image/jpeg',
  [IMAGE_EXTENSIONS.WEBP]: 'image/webp',
  [IMAGE_EXTENSIONS.GIF]: 'image/gif',
};

export const dataURItoBlob = (dataURI: string): Blob => {
  // convert base64 to raw binary data held in a string
  const byteString = atob(dataURI.split(',')[1]);

  // separate out the mime component
  const mimeString = dataURI.split(',')[0].split(':')[1].split(';')[0];

  // write the bytes of the string to an ArrayBuffer
  const arrayBuffer = new ArrayBuffer(byteString.length);
  const _ia = new Uint8Array(arrayBuffer);

  for (let i = 0; i < byteString.length; i++) {
    _ia[i] = byteString.charCodeAt(i);
  }

  const dataView = new DataView(arrayBuffer);
  const blob = new Blob([dataView], { type: mimeString });

  return blob;
};

/**
 * Create a blob by image url
 */
export const createImageBlobByUrl = (url: string, mimeType: string): Promise<Blob> => {
  return new Promise((resolve, reject) => {
    const image = new Image();

    image.src = url;

    image.onload = () => {
      const { width, height } = image;
      const canvas = Object.assign(document.createElement('canvas'), { width, height });
      const context = canvas.getContext('2d');

      if (context !== null) {
        context.fillStyle = 'white';
        context.fillRect(0, 0, canvas.width, canvas.height);
        context.drawImage(image, 0, 0);

        const dataURI = canvas.toDataURL(mimeType);
        const blob = dataURItoBlob(dataURI);

        resolve(blob);
      } else {
        reject(new Error(''));
      }
    };

    image.onerror = (error) => reject(error);
  });
};

export const createSvgBlob = (svgElement: SVGSVGElement): Blob => {
  const data = new XMLSerializer().serializeToString(svgElement);
  const blob = new Blob([data], { type: 'image/svg+xml' });

  return blob;
};

export const svgSaveAs = async (
  svgElement: SVGSVGElement,
  name: string,
  extension: IMAGE_EXTENSIONS = IMAGE_EXTENSIONS.SVG
): Promise<void> => {
  let blob = createSvgBlob(svgElement);

  if (extension !== IMAGE_EXTENSIONS.SVG) {
    const mimeType = IMAGE_MIME_TYPES[extension];
    const url = URL.createObjectURL(blob);

    blob = await createImageBlobByUrl(url, mimeType);

    URL.revokeObjectURL(url);
  }

  const filename = `${name}${extension}`;

  saveAs(blob, filename);
};

export function getBase64Icon(icon: string): string {
  const BASE64_PNG_PREFIX = 'data:image/png;base64';
  const XML_SVG_PREFIX = 'data:image/svg+xml';

  if (icon.startsWith(BASE64_PNG_PREFIX)) return icon;
  if (icon.startsWith(XML_SVG_PREFIX)) {
    // Take svg string starting from '<' char up to end
    const svgUriEncoded = icon.substring(icon.indexOf('%3C'));
    const svgUriDecoded = decodeURIComponent(svgUriEncoded);
    const base64SvgEncoded = base64.encode(svgUriDecoded);
    return `${XML_SVG_PREFIX};base64,${base64SvgEncoded}`;
  }

  return '';
}
