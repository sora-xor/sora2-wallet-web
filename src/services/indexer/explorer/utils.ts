import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
