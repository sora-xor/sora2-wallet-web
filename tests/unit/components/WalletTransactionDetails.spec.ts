import Vuex from 'vuex'
import { shallowMount } from '@vue/test-utils'
import { History } from '@sora-substrate/util'

import { useDescribe, localVue } from '../../utils'
import { MOCK_ACCOUNT_ASSETS, MOCK_HISTORY } from '../../utils/mock'

import WalletTransactionDetails from '@/components/WalletTransactionDetails.vue'

const createStore = (tx: History) => new Vuex.Store({
  modules: {
    Account: {
      getters: {
        selectedTransaction: () => tx,
        accountAssets: () => MOCK_ACCOUNT_ASSETS
      },
      actions: {
        getTransactionDetails: jest.fn(),
        getAccountActivity: jest.fn()
      }
    },
    Router: {
      getters: {
        currentRouteParams: () => ({ id: '1', asset: MOCK_ACCOUNT_ASSETS[0] })
      },
      actions: {
        navigate: jest.fn()
      }
    }
  } as any
})

useDescribe('WalletTransactionDetails.vue', WalletTransactionDetails, () => {
  MOCK_HISTORY.map(item => it(`[${item.type}, ${item.status}]: should be rendered correctly`, () => {
    const wrapper = shallowMount(WalletTransactionDetails, { localVue, store: createStore(item) })
    expect(wrapper.element).toMatchSnapshot()
  }))
})
