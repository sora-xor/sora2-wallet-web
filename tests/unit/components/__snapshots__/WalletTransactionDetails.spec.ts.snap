// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletTransactionDetails.vue [AddLiquidity, finalized]: should be rendered correctly 1`] = `
<wallet-base-stub
  actionicon=""
  actiontooltip=""
  showback="true"
  title=""
  tooltip=""
>
  <div
    class="transaction"
  >
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="0xB8c77482e4...52C74426C6DD"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/block/0xB8c77482e45F1F4d123DeRwQ5F52C74426C6DD"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        <span
          class="transaction-status"
        >
          
        </span>
      </div>
       
      <s-icon-stub
        name="basic-check-mark-24"
      />
    </div>
     
    <!---->
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        27.11.2020, 11:59:59
      </div>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        0.1 XOR
      </div>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        0.9 PSWAP
      </div>
    </div>
     
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="5HVmWWpBi69c...pveRnfozNg5K"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/account/5HVmWWpBi69cmmDqWE4R6yxxJ2pveRnfozNg5K"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <!---->
  </div>
</wallet-base-stub>
`;

exports[`WalletTransactionDetails.vue [RegisterAsset, finalized]: should be rendered correctly 1`] = `
<wallet-base-stub
  actionicon=""
  actiontooltip=""
  showback="true"
  title=""
  tooltip=""
>
  <div
    class="transaction"
  >
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="0xB8c77482e4...52C74426C6DD"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/block/0xB8c77482e45F1F4d123DeRwQ5F52C74426C6DD"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        <span
          class="transaction-status"
        >
          
        </span>
      </div>
       
      <s-icon-stub
        name="basic-check-mark-24"
      />
    </div>
     
    <!---->
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        27.11.2020, 11:59:59
      </div>
    </div>
     
    <!---->
     
    <!---->
     
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="5HVmWWpBi69c...pveRnfozNg5K"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/account/5HVmWWpBi69cmmDqWE4R6yxxJ2pveRnfozNg5K"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <!---->
  </div>
</wallet-base-stub>
`;

exports[`WalletTransactionDetails.vue [RemoveLiquidity, finalized]: should be rendered correctly 1`] = `
<wallet-base-stub
  actionicon=""
  actiontooltip=""
  showback="true"
  title=""
  tooltip=""
>
  <div
    class="transaction"
  >
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="0xB8c77482e4...52C74426C6DD"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/block/0xB8c77482e45F1F4d123DeRwQ5F52C74426C6DD"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        <span
          class="transaction-status"
        >
          
        </span>
      </div>
       
      <s-icon-stub
        name="basic-check-mark-24"
      />
    </div>
     
    <!---->
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        27.11.2020, 11:59:59
      </div>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        100.123 XOR undefined PSWAP
      </div>
    </div>
     
    <!---->
     
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="5HVmWWpBi69c...pveRnfozNg5K"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/account/5HVmWWpBi69cmmDqWE4R6yxxJ2pveRnfozNg5K"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <!---->
  </div>
</wallet-base-stub>
`;

exports[`WalletTransactionDetails.vue [Swap, error]: should be rendered correctly 1`] = `
<wallet-base-stub
  actionicon=""
  actiontooltip=""
  showback="true"
  title=""
  tooltip=""
>
  <div
    class="transaction"
  >
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="0xB8c77482e4...52C74426C6DD"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/block/0xB8c77482e45F1F4d123DeRwQ5F52C74426C6DD"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        <span
          class="transaction-status transaction-status--error"
        >
          
        </span>
      </div>
       
      <!---->
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        <span
          class="transaction-status transaction-status--error"
        >
          Test error message!
        </span>
      </div>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        27.11.2020, 11:59:59
      </div>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        0.1 XOR
      </div>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        0.9 PSWAP
      </div>
    </div>
     
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="5HVmWWpBi69c...pveRnfozNg5K"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/account/5HVmWWpBi69cmmDqWE4R6yxxJ2pveRnfozNg5K"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <!---->
  </div>
</wallet-base-stub>
`;

exports[`WalletTransactionDetails.vue [Transfer, finalized]: should be rendered correctly 1`] = `
<wallet-base-stub
  actionicon=""
  actiontooltip=""
  showback="true"
  title=""
  tooltip=""
>
  <div
    class="transaction"
  >
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="0xB8c77482e4...52C74426C6D7"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/transaction/0xB8c77482e45F1F4d123DeRwQ5F52C74426C6D7"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        <span
          class="transaction-status"
        >
          
        </span>
      </div>
       
      <s-icon-stub
        name="basic-check-mark-24"
      />
    </div>
     
    <!---->
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        27.11.2020, 11:59:59
      </div>
    </div>
     
    <div
      class="transaction-row s-flex"
    >
      <div
        class="transaction-row_key"
      >
        
        
      
      </div>
       
      <div
        class="transaction-row_value"
      >
        0.1 XOR
      </div>
    </div>
     
    <!---->
     
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="5HVmWWpBi69c...pveRnfozNg5K"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/account/5HVmWWpBi69cmmDqWE4R6yxxJ2pveRnfozNg5K"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
     
    <div
      class="s-input-container"
    >
      <s-input-stub
        accept=""
        autocomplete="off"
        borderradius="small"
        form=""
        label=""
        name=""
        placeholder=""
        prefix=""
        readonly="true"
        size="big"
        suffix=""
        tabindex=""
        type="text"
        usedesignsystem="true"
        value="0xB8c77482e4...52C74426C6DD"
      />
       
      <s-button-stub
        alternative="true"
        borderradius="small"
        class="s-button--copy"
        icon="basic-copy-24"
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="action"
        usedesignsystem="true"
      />
       
      <s-dropdown-stub
        borderradius="mini"
        buttontype="secondary"
        class="s-dropdown-menu"
        hideonclick="true"
        hidetimeout="150"
        icon="basic-more-vertical-24"
        placement="bottom-end"
        showtimeout="250"
        size="medium"
        tabindex="0"
        trigger="hover"
        type="ellipsis"
      >
        <template>
          <a
            class="transaction-link"
            href="https://explorer.s2.dev.sora2.soramitsu.co.jp/sora-dev/account/0xB8c77482e45F1F4d123DeRwQ5F52C74426C6DD"
            rel="nofollow noopener"
            target="_blank"
          >
            <s-dropdown-item-stub
              class="s-dropdown-menu__item"
            >
              
              
            
            </s-dropdown-item-stub>
          </a>
        </template>
      </s-dropdown-stub>
    </div>
  </div>
</wallet-base-stub>
`;
